client/scripts/controllers/server/socket/chainEventsNs.ts

```diff
@@ -0,0 +1,68 @@
+import { WebsocketMessageType, WebsocketNamespaces } from 'types';
+import app from 'state';
+import { Notification, NotificationSubscription } from 'models';
+import { io } from 'socket.io-client';
+
+export class ChainEventsNamespace {
+  private ceNs;
+  private _isConnected = false;
+
+  constructor() {
+    this.ceNs = io(`/${WebsocketNamespaces.ChainEvents}`, {
+      transports: ['websocket'],
+    });
+    this.ceNs.on('connect', this.onconnect.bind(this));
+    this.ceNs.on('disconnect', this.ondisconnect.bind(this));
+    this.ceNs.on(
+      WebsocketMessageType.ChainEventNotification,
+      this.onChainEvent.bind(this)
+    );
+  }
+
+  public addChainEventSubscriptions(subs: NotificationSubscription[]) {
+    if (this._isConnected) {
+      const eventTypes = subs.map((x) => x.ChainEventType?.id).filter((x) => !!x);
+      console.log('Adding Websocket subscriptions for:', eventTypes);
+      this.ceNs.emit('newSubscriptions', eventTypes);
+    } else {
+      console.log('ChainEventsNamespace is not connected');
+    }
+  }
+
+  public deleteChainEventSubscriptions(subs: NotificationSubscription[]) {
+    if (this._isConnected) {
+      const eventTypes = subs.map((x) => x.ChainEventType?.id).filter((x) => !!x);
+      console.log('Deleting Websocket subscriptions for:', eventTypes);
+      this.ceNs.emit(
+        'deleteSubscriptions',
+        subs.map((x) => x.ChainEventType?.id)
+      );
+    } else {
+      console.log('ChainEventsNamespace is not connected');
+    }
+  }
+
+  private onChainEvent(notification: any) {
+    const subscription = app.user.notifications.subscriptions.find(
+      (sub) => sub.ChainEventType?.id === notification.ChainEvent.ChainEventType.id
```

Comment:
Any way we can make subscriptions an object so we can index the event directly rather than doing a find?

Reply:
That should probably be another PR to generally optimize notifications as subscriptions cover all notifications so it would require changes in places irrelevant to websockets/chain-events. I don't mind including it here if you feel it's necessary though.